packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -108,64 +118,122 @@ export function memoizeTracked<T>(
+
+      cache = debugCache;
+    }
+
+    let { tag, snapshot, fn } = cache;
+
+    if (tag === undefined || !validateTag(tag, snapshot)) {
       beginTrackFrame();
 
       try {
         if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
+          runInAutotrackingTransaction!(() => (cache.lastValue = fn(...args)), debuggingContext!);
         } else {
-          lastValue = callback(...args);
+          cache.lastValue = fn(...args);
         }
       } finally {
         tag = endTrackFrame();
-        snapshot = valueForTag(tag);
+        cache.tag = tag;
+        cache.snapshot = valueForTag(tag);
         consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
       }
     } else {
       consumeTag(tag);
     }
 
-    return lastValue!;
-  };
-
-  if (DEBUG) {
-    IS_CONST_MAP.set(memoized, undefined);
+    return cache.lastValue!;
   }
 
-  return memoized;
-}
+  static isConst(cache: Cache) {
+    if (DEBUG) {
+      let debugCache = (cache as any)[DEBUG_CACHE];
+
+      if (!(debugCache instanceof Cache)) {
+        throw new Error(
+          `isConst() can only be used on an instance of a cache created with createCache(). Called with: ${String(
+            cache
+          )}`
+        );
+      }
 
-export function isConstMemo(memoized: Function) {
-  if (DEBUG && IS_CONST_MAP.has(memoized) && IS_CONST_MAP.get(memoized) === undefined) {
-    throw new Error(
-      'Attempted to call `isConstMemo` on a memoized function, but the function has not been run at least once yet. You cannot know if a memoized function is constant or not until it has been run at least once. Call the function, then pass it to `isConstMemo`.'
-    );
+      if (debugCache.tag === undefined) {
+        throw new Error(
+          `isConst() can only be used on a cache once getValue() has been called at least once. Called with cache function:\n\n${String(
+            debugCache.fn
+          )}`
+        );
+      }
+
+      cache = debugCache;
+    }
+
+    return isConstTag(cache.tag!);
   }
+}
 
-  return IS_CONST_MAP.get(memoized) === true;
+export function createCache<T>(fn: () => T): Cache<T> {
+  return new Cache(fn);
 }
 
+export const getValue = Cache.getValue;
+
+export const isConst = Cache.isConst;
+
 //////////
 
+// Legacy tracking APIs
```

Comment:
Can you define what legacy means in this context? Maybe even make an issue to remove them and link to that from here (that way we know when we can remove the code).

Reply:
Yeah, I'll add a comment. `track()` shouldn't be necessary at all in the VM once the autotracking refactors are merged, and we should generally be moving away from it. It may be necessary in Ember for a while longer, but I think we'll be able to drop it in favor of `cache` sooner rather than later.

`untrack` is currently mainly used to handle places that were previously not tracked, and that tracking now would cause backtracking rerender assertions. I think once we move everyone forward onto modern APIs, we'll probably be able to remove it, but I'm not sure yet.

Reply:
Sounds good. I mostly want to make sure the code comments are actionable. This one infers something is going away, but it would be pretty hard to be able to tell if it were ok to remove if you came back and read this again in ~ 6months...